### PR TITLE
Auto mapping feature ids from old shorthand syntax to devcontainers/features

### PR DESCRIPTION
**Pull request by samruddhikhandale**
Addresses - https://github.com/devcontainers/cli/issues/88 and https://github.com/devcontainers/cli/issues/88

Mapping of old shorthand syntax of features
- Migrated feature ids are mapped to `ghcr.io/devcontainers/features/${id}:1`
- Renamed feature ids are mapped to `ghcr.io/devcontainers/features/${id}:1`
- Deprecated feature ids are kept as it is

Added tests? 
Yes! 

**Dependency -** 
Should not be merged until we have started pointing people to new repository and pulled off the plug from https://github.com/microsoft/vscode-dev-containers

### CODE REVIEW

File: src/spec-configuration/containerFeaturesConfiguration.ts [hunk 1]
```diff
@@ -505,12 +505,38 @@ export async function getFeatureIdType(output: Log, env: NodeJS.ProcessEnv, id:
 	return { type: 'github-repo', manifest: undefined };
 }
 
+export function getBackwardCompatibleFeatureId(id: string) {
```
Comment (samruddhikhandale) [inline src/spec-configuration/containerFeaturesConfiguration.ts:508]: 💀 Should not be merged until we are ready for open sourcing /images and /features repositories

File: src/spec-configuration/containerFeaturesConfiguration.ts [hunk 2]
```diff
@@ -514,14 +514,42 @@ export async function getFeatureIdType(output: Log, env: NodeJS.ProcessEnv, user
...
+	renamedFeatures.set('golang', 'go');
+	renamedFeatures.set('common', 'common-utils');
+
+	const newFeaturePath = 'ghcr.io/devcontainers/features';
+	// Note: Pin the versionBackwardComp to '1' to avoid breaking changes.
+	const versionBackwardComp = '1';
+
+	// Mapping feature references (old shorthand syntax) from "microsoft/vscode-dev-containers" to "ghcr.io/devcontainers/features"
+	if (migratedfeatures.includes(id)) {
+		return `${newFeaturePath}/${id}:${versionBackwardComp}`;
```
Comment (chrmarti) [inline src/spec-configuration/containerFeaturesConfiguration.ts:529]: We are using `@` as the separator (unless this changed, but the parsing below still uses that).

File: src/spec-configuration/containerFeaturesConfiguration.ts [hunk 3]
```diff
@@ -514,14 +514,42 @@ export async function getFeatureIdType(output: Log, env: NodeJS.ProcessEnv, user
...
+	}
+
+	// Mapping feature references (renamed old shorthand syntax) from "microsoft/vscode-dev-containers" to "ghcr.io/devcontainers/features"
+	if (renamedFeatures.get(id) !== undefined) {
+		return `${newFeaturePath}/${renamedFeatures.get(id)}:${versionBackwardComp}`;
```
Comment (chrmarti) [inline src/spec-configuration/containerFeaturesConfiguration.ts:534]: Use `@` as separator before version.

File: src/spec-configuration/containerFeaturesConfiguration.ts [hunk 4]
```diff
@@ -514,14 +514,42 @@ export async function getFeatureIdType(output: Log, env: NodeJS.ProcessEnv, user
 	}
 }
 
+export function getBackwardCompatibleFeatureId(id: string) {
+	const migratedfeatures = ['aws-cli', 'azure-cli', 'desktop-lite', 'docker-in-docker', 'docker-from-docker', 'dotnet', 'git', 'git-lfs', 'github-cli', 'java', 'kubectl-helm-minikube', 'node', 'powershell', 'python', 'ruby', 'rust', 'sshd', 'terraform'];
```
Comment (chrmarti) [inline src/spec-configuration/containerFeaturesConfiguration.ts:518]: I see fish, gradle, homebrew, jupyterlab, maven are missing, assuming these are deprecated, should we add a log message when one of these is used?

File: src/spec-configuration/containerFeaturesConfiguration.ts (hunk 2 above)
Comment (samruddhikhandale) [reply src/spec-configuration/containerFeaturesConfiguration.ts:529]: According the [spec](https://github.com/devcontainers/spec/blob/main/proposals/devcontainer-features.md#referencing-a-feature) changes for OCI, we now use `:` 😄 

> `<oci-registry>/<namespace>/<feature>[:<semantic-version>]`
> eg. ghcr.io/user/repo/go:1

File: src/spec-configuration/containerFeaturesConfiguration.ts (hunk 3 above)
Comment (samruddhikhandale) [reply src/spec-configuration/containerFeaturesConfiguration.ts:534]: Same comment as above!

File: src/spec-configuration/containerFeaturesConfiguration.ts (hunk 2 above)
Comment (joshspicer) [reply src/spec-configuration/containerFeaturesConfiguration.ts:529]: We're trying to conform to the OCI spec for identifiers (you should be to just do an `oras pull <identifier>`).  

Tags are preceded by a `:`.  One thing we _don't_ support right now is `...@<digest>`, which we should do and update the spec accordingly.

File: src/spec-configuration/containerFeaturesConfiguration.ts (hunk 4 above)
Comment (samruddhikhandale) [reply src/spec-configuration/containerFeaturesConfiguration.ts:518]: Sounds good. Added a message. 
Once features repo is open source, I'm thinking of adding more detailed log messages for these as well as migrated/renamed features

File: src/spec-configuration/containerFeaturesConfiguration.ts (hunk 2 above)
Comment (chrmarti) [reply src/spec-configuration/containerFeaturesConfiguration.ts:530]: Sounds good, thanks.

**Review by chrmarti — APPROVED** [review]

**Review by chrmarti — CHANGES_REQUESTED** [review]
Removing approval to discuss the option to let the new functionality get more testing before we redirect existing configurations to use it.

**Review by chrmarti — APPROVED** [review]
Reapproving since Remote-Containers stable version 0.245.0 is released. (As discussed.)

File: src/spec-configuration/containerFeaturesConfiguration.ts (hunk 4 above)
Comment (samruddhikhandale) [reply src/spec-configuration/containerFeaturesConfiguration.ts:519]: I removed the warning log messages, as per today's discussion. 

I'll add them back only when auto-mapping is ready to be put in Remote-Containers STABLE version

@chrmarti  does that sound correct to you?

Comment (chrmarti) [reply src/spec-configuration/containerFeaturesConfiguration.ts:519]: Makes sense, thanks!

**Review by chrmarti — APPROVED** [review]

**Pull request merged**